daemon/src/model.rs

```diff
@@ -1,62 +1,81 @@
 use bdk::bitcoin::{Address, Amount};
 use reqwest::Url;
-use rust_decimal::prelude::ToPrimitive;
+use rust_decimal::prelude::{FromPrimitive, ToPrimitive};
 use rust_decimal::Decimal;
 use rust_decimal_macros::dec;
 use serde::{Deserialize, Serialize};
 use serde_with::{DeserializeFromStr, SerializeDisplay};
+use std::ops::{Add, Div, Mul, Sub};
 use std::time::SystemTime;
 use std::{fmt, str};
 use time::{OffsetDateTime, PrimitiveDateTime, Time};
 use uuid::Uuid;
 
 pub mod cfd;
 
-#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
-pub struct Usd(Decimal);
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Exchange rate of zero is not allowed.")]
+    ZeroExchangeRate,
+    #[error("Negative exchange rate is not allowed.")]
+    NegativeExchangeRate,
+    #[error("Leverage must be >= 1.")]
+    NonPositiveLeverage,
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, PartialOrd)]
+pub struct Usd {
+    value: Decimal,
+}
 
 impl Usd {
-    pub fn new(dec: Decimal) -> Self {
-        Self(dec)
+    pub fn new(value: Decimal) -> Self {
+        Usd { value }
     }
 
-    pub fn checked_add(&self, other: Usd) -> Result<Usd> {
-        let result = self.0.checked_add(other.0).context("addition error")?;
-        Ok(Usd(result))
+    pub fn try_into_u64(&self) -> Result<u64> {
+        self.value
+            .to_u64()
+            .context("could not fit decimal into u64")
     }
+}
 
-    pub fn checked_sub(&self, other: Usd) -> Result<Usd> {
-        let result = self.0.checked_sub(other.0).context("subtraction error")?;
-        Ok(Usd(result))
-    }
+// impl From<f64> for Result<Usd> {
+//     fn from(value: f64) -> Result<Self> {
+//         let decimal = Decimal::from_f64(value).context("Cannot convert value to decimal")?;
+//         Ok(Usd::new(decimal))
+//     }
+// }
 
-    // TODO: Usd * Usd = Usd^2 not Usd !!!
-    pub fn checked_mul(&self, other: Usd) -> Result<Usd> {
-        let result = self
-            .0
-            .checked_mul(other.0)
-            .context("multiplication error")?;
-        Ok(Usd(result))
+impl fmt::Display for Usd {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.value.fmt(f)
     }
+}
 
-    pub fn checked_div(&self, other: Usd) -> Result<Usd> {
-        let result = self.0.checked_div(other.0).context("division error")?;
-        Ok(Usd(result))
-    }
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, PartialOrd)]
+pub struct ExchangeRate {
+    value: Decimal,
+}
 
-    pub fn try_into_u64(&self) -> Result<u64> {
-        self.0.to_u64().context("could not fit decimal into u64")
-    }
+impl ExchangeRate {
+    pub fn new(value: Decimal) -> Result<Self, Error> {
+        if value.to_f64().unwrap().abs() <= f64::EPSILON {
+            return Result::Err(Error::ZeroExchangeRate);
+        } else if value.to_f64().unwrap() < 0f64 {
+            return Result::Err(Error::NegativeExchangeRate);
+        };
```

Comment:
Just a style thing but I don't like `else if` blocks. I would change this to:

```suggestion
        if value.to_f64().unwrap().abs() <= f64::EPSILON {
            return Result::Err(Error::ZeroExchangeRate);
        }
        
        if value.to_f64().unwrap() < 0f64 {
            return Result::Err(Error::NegativeExchangeRate);
        };
```